---
title: "seurat_pbmc_workflow"
author: "Itzel Espinoza"
date: "2024-06-22"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
```



```{r, echo=FALSE}

# The dataset of the Peripheral Blood Mononuclear Cells (PBMC) is available at  https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz

# ------ Loading the PBMC dataset ------
# pbmc_data = represents the raw count matrix where rows correspond to genes (or features), and columns correspond to individual cells

pbmc_data <- Read10X(data.dir = "../filtered_gene_bc_matrices/hg19/")

# ------ Initializing the Seurat object with the raw (non-normalized data) ------
# Output is a container that holds single-cell RNA sequencing data, along with associated metadata and analysis results
# min.cells = threshold for the min. number of cells in which a gene must be detected to be included in my analysis, which filter out lowly expressed genes that may be noise
# min.features = threshold for the min. number of genes (or features) that must be detected in a given cell for that cell to be included, which excludes empty or poor-quality cells that don't have enough data

pbmc <- CreateSeuratObject(counts = pbmc_data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc

# ------ Exploring the count matrix ------ 
# Taking a look at what genes are inside the count matrix
rownames(pbmc_data)

# Genes in the first twenty cells
pbmc_data[c("TP53","CASP9","MAP3K6","MT-CO3"), 1:20]

# ------ Memory Usage and Efficiency  ------ 

# Sparse-matrix
# The . values in the matrix represent 0s (no molecules detected) and Seurat uses a sparse-matrix representation whenever possible for more memory and speed savings
sparse.size <- object.size(pbmc_data)
sparse.size

# Comparison to a Regular matrix (x23.7 bigger)
memory.size <- object.size(as.matrix(pbmc_data))
memory.size
```

# Visualizing Quality Control Metrics with Violin Plots

- `nFeature_RNA`: The number of genes detected in each cell.
- `nCount_RNA`: The total number of RNA molecules detected in each cell (overall count).
- `percent.mt`: The percentage of mitochondrial gene expression in each cell

```{r}

# ------ Quality Control  ------ 
# Identify mitochondrial genes, these genes start with "MT-"
# Low-quality / dying cells often show extensive mitochondrial contamination, will have to filter out later
mt_rows <- grep("^MT-", rownames(pbmc_data))
filtered_matrix <- pbmc_data[mt_rows, ]


#  Adding Mitochondrial Gene QC Metric
# This computes the % of mitochondrial gene expression for each cell and stores it as a metadata column (percent.mt) in the pbmc Seurat object
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

#  ------ QC metrics as a violin plot ------ 
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```

# Visualizing Feature-Feature Relationships

## Scatter Plot: nCount_RNA vs. percent.mt

This figure shows the total RNA count per cell `(nCount_RNA)` against the percentage of mitochondrial genes expressed `(percent.mt)`

*Interpretation:*
- Typically, a higher percentage of mitochondrial gene expression is associated with lower total RNA counts. This pattern is often observed in stressed or dying cells, where mitochondrial RNA levels increase as the cell deteriorates, leading to a decrease in overall RNA content. Conversely, healthy cells usually maintain a higher total RNA count with a lower percentage of mitochondrial gene expression.
- A correlation coefficient of -0.13 between nCount_RNA (total RNA count per cell) and percent.mt (percentage of mitochondrial gene expression) suggests a very weak negative relationship between these two variables.
- Overall, this figure shows that cells with higher total RNA counts do not significantly differ in their mitochondrial gene expression percentages compared to cells with lower RNA counts.

```{r}

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt") + geom_point(color = "royalblue") + NoLegend()
plot1
```


## Scatter Plot: nCount_RNA vs. nFeature_RNA

This plots the total RNA count per cell (nCount_RNA) against the number of features detected per cell (nFeature_RNA).

*Interpretation:*
- A strong correlation suggests that cells with higher RNA content also express a greater number of genes, which is typical in healthy cells. Deviations from this pattern may indicate issues like doublets or low-quality cells.
- This figure shows a strong positive correlation with a correlation coefficient of `0.95`, meaning that cells with higher total RNA counts tend to express a greater number of genes.

```{r}
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot2

```


## Side by Side Comparison of Scatter Plots
```{r}
plot1 + plot2

```

# Quality Control and Cell Filtering

- *Low Gene Detection* `nFeature_RNA > 200`: Includes cells that have detected more than 200 genes.
- *High Gene Detection* `nFeature_RNA < 2500`: Includes cells that have detected fewer than 2,500 genes.
- *High Mitochondrial Content* `percent.mt < 5`: Includes cells where less than 5% of the detected genes are mitochondrial genes.

```{r}

# Filtering cells in the Seurat object based on specific quality control (QC) metrics:
# 13714 features across 2638 samples ( -62 samples after filtering)

pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```


# Data Normalization

Normalizing gene expression counts across all cells in the dataset. Normalizing corrects for differences in sequencing depth, makes gene expression comparable across cells and prepares data for downstream analysis such as dimensionality reduction & clustering. 

*Log Normalization*
- Global-scaling normalization is useful here because it adjusts each cell's gene expression values based on total RNA content, making them comparable across all cells.
- Log transformation also stabilizes variance, reducing the influence of highly expressed genes
- Prepares for downstream statistical analysis by helping approximate a normal distribution

*The global-scaling normalization assumes that:*
Most genes are not differentially expressed – The total RNA count per cell reflects technical effects rather than biological differences.

*Potential issue:* If some cell types have inherently higher RNA content (e.g., due to size differences), global-scaling normalization might artificially reduce the expression of highly transcribed genes in those cells.

```{r}
# ------ Standard method of normalization ------ 
# The "LogNormalize" method follows these steps:
# 1. For each gene in each cell:
# 2. Divide the gene’s expression count by the total expression of that cell.
# 3. Multiply the result by a scale factor (default: 10,000) to avoid very small numbers.
# 4. Apply a log transformation (log1p, or log(x + 1)) to stabilize variance.

pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)

pbmc <- NormalizeData(pbmc)

# Alternative normalization method: To use SCTransform as this replaces the need to run NormalizeData, FindVariableFeatures, or ScaleData
```

# Identification of highly variable features

Identifying highly variable features (genes) is a critical pre-processing step before dimensionality reduction and clustering. The goal is to retain genes that exhibit meaningful biological variation while filtering out noise and technical artifacts.

## Scatter plot: All variable genes
```{r}
# ------ STEP 1 ------ 
# This function identifies genes that show high variance across cells, which helps capture biological heterogeneity.

pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# `vst` uses the variance-stabilizing transformation (VST) method, which models the mean-variance relationship in single-cell data
# `nfeatures = 2000` Selects the top 2,000 most variable genes for downstream analysis

# ------ STEP 2 ------ 
# Extracting and examining the Top Variable Genes
# This function retrieves the genes identified as highly variable

top10 <- head(VariableFeatures(pbmc), 10)

# ------ STEP 3 ------ 
# Visualizing Variable Genes
# This function generates a scatter plot of all genes, showing their variability
# Helps confirm that the feature selection method captures genes with high variance

plot1 <- VariableFeaturePlot(pbmc, cols = c("black", "darkorange2")) +
  ggtitle("Highly Variable Genes in PBMC Dataset")
plot1


```

## Scatter plot: Highlighting the Top 10 Variable Genes

The top 10 highly variable genes identified in this figure are primarily involved in immune response, inflammation, and cell signaling, suggesting substantial heterogeneity within the immune cell populations in the PBMC dataset. These genes, including markers for neutrophils, macrophages, and B cells, can provide valuable insights into immune activation and could serve as potential biomarkers for immune-related diseases or conditions.

*Biological Significance:*
- These genes are involved in immune and inflammatory responses
- Many of them are markers of inflammatory immune cells (e.g., S100A9, S100A8, LYZ, GNLY) or iron metabolism (e.g., FTL, FTH1)
- Platelet activation genes like PPBP and PF4 indicate a potential role in vascular and immune interactions
```{r}
# ------ STEP 3 ------ 
# Visualizing Variable Genes
# This function labels the top 10 most variable genes on the plot to highlight them

plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE, size = 4) +
  ggtitle("Top 10 Highly Variable Genes in PBMC Dataset")
plot2

# 11,714 = number of genes that did not show significant variation across the cells in your dataset
# 2,000 = number of variable genes considered to be the most informative for capturing biological heterogeneity within the dataset
```

## Scatter plot: Side by side
```{r}
plot3 <- plot1 + plot2
plot3
```


